### ISSUE TITLE
syncy is unusable with node_modules/ in destination

### ISSUE THREAD

**Issue by eisbehr-** [mrmlnc/syncy#37]
I wanted to use `syncy` to copy development files _on-the-fly_ to a test system when a files got changed with `gulp`. So far no problem at all, workes great. But whenever I install all the modules on the other system, `syncy` takes a huge amount of time to sync (about 2 minutes for 14 files).

I have tried to use `ignoreInDest` option with `node_mocules/**` glob, but this makes no difference. It is right, the folder will be ignored and not deleted, but it seems `syncy` still looks for all files in it. 

I use a simple `gulp.watch` task to execute the sync. As said, workes great in general, until the modules are installed in the destination. Here is how my task looks like:

```JS
gulp.task("sync-watch", function() {
    gulp.watch(["./*.js", "./*.json"], function() {
        syncy(["./*.js", "./*.json"], '\\\\127.0.0.1\\Share\\folder\\', {
            ignoreInDest: "node_modules/**"
        })
        .then(function() {
            util.log("Done with './'");
        })
        .catch(util.log);
    });
});
```

**Comment by eisbehr-** [comment]
I stepped a bit into your code, to see what is so time-consuming there, and it seems that `globby` takes a long time. When I exclude `node_modules/` at the `globby` call inside your `run` function, everything is working fast again.

```JS
const destFiles = await globby(['**', "!node_modules/**"], <glob.IOptions>{
    cwd: dest,
    dot: true,
    nosort: true
});
```

Maybe use the value of `ignoreInDest` to exclude this on this `globby` call? Or maybe add another option for this. For now I use the changes version of `syncy` to use it on my project ...

**Comment by mrmlnc** [comment]
Please, give me a few details:

  * OS?
  * HDD or SSD?
  * How many files in the directory?
  * How quickly files are copied if you use native method?

Now it sounds like wrong usage. You can use `["*.js", "!**/node_modules/**"]`. 

Maybe I'm wrong — I'll take a look tomorrow in more detail.

**Comment by eisbehr-** [comment]
I don't have the details here. Will give them to you these days. Even if I don't understand, why the native time is a factor here, when everything workes fine without modules installed in destination...  But:

`*.js` don't include `node_modules/` folder! It's just files in the root directory, not sub-directories. So there is no need to exclude `node_modules/`. You would be correct when I had used `**/*.js`.

**Comment by eisbehr-** [comment]
And to be clear, if it was confusing:

My issue is just about `node_modules/` in **destination** folder, when I installed it there. I have no problem with `node_modules/` in source folder and I **don't** copy the folder to destination too! The problem is, that `syncy` read in all folders in destination even when it's not necessary. This consumes unneeded time, because the information is not needed imo. When there is a huge amount of files this takes even longer. 

On a local system with fast drives that would probably not even a problem. But in my case it's a device connected over VPN with a samba share. So it's not as fast and so `globby` takes even more time to read in the data ...

**Comment by mrmlnc** [comment]
@eisbehr-,

Yeap, I understand your problem, but now I don't have a solution. This behavior applies to the algorithm.

Your [temporary solution](https://github.com/mrmlnc/syncy/issues/37#issuecomment-282325783) will work not correctly. See https://github.com/mrmlnc/syncy/issues/26 for more details.

I agree that the current algorithm has performance problems. I will think about what I can do.

**Comment by eisbehr-** [comment]
Im not as deep in the code `syncy`, but does this really infect the behavior of #26? If already removed from the `globby` call, there is no other handling of the files, even no delete action. They are completly ignored. Or is there anywhere a second read-in of the files in destination?

I have not noticed any problems with my solution till now. There are no deleted files in destination. I'll now remove all ignored files automatically from `globby`.

```JS
// Get files from destination directory
const ignoreInDestArr = Array.isArray(options.ignoreInDest) ? options.ignoreInDest : [options.ignoreInDest];
const destFiles = await globby(['**'].concat(ignoreInDestArr.map(v => { return "!" + v; }), <glob.IOptions>{
    cwd: dest,
    dot: true,
    nosort: true
});
```

**Comment by eisbehr-** [comment]
Hey! Still no idea how to improve this? ;)
Would still love to see a change here! ❤️

**Comment by mrmlnc** [comment]
Hi, @eisbehr-,

I create before #41 and thought about the algorithm, but have no free time now 😢

**Comment by fourthmeal70** [comment]
I'm wondering if this issue has been given any more attention? I'm having a similar issue, where once I added a directory with many files in it to the "ignoreInDest" option, the sync takes a very long time(almost 7 hours on the last run). This directory has almost 100k image files in it with various levels of subdirectories. If I deploy to a different destination that does not have that or other large directories I want to ignore, the sync runs in just a few minutes. 

It must be processing each file in the ignored directory for some reason. I think if a directory is in the ignore list, then it should skip it all together and not go into it at all. Otherwise, I this is a great tool, hope this get's some additional attention!

**Comment by eisbehr-** [comment]
Yes @fourthmeal70 , I think it has just no priority here. Or project is dead. idk :/

But you may want to try my change posted above, and see if that works for you too and improved the time consumption. It's just a code change on the source. Maybe if this works for you, he will use it in the live version too.

Just change the `globby` call in [line 70](https://github.com/mrmlnc/syncy/blob/master/src/syncy.ts#L70) like this in the source:

```JS
// Get files from destination directory
const ignoreInDestArr = Array.isArray(options.ignoreInDest) ? options.ignoreInDest : [options.ignoreInDest];
const destFiles = await globby(['**'].concat(ignoreInDestArr.map(v => { return "!" + v; }), <glob.IOptions>{
    cwd: dest,
    dot: true,
    nosort: true
});
```

**Comment by mrmlnc** [comment]
Hello, @eisbehr- and @fourthmeal70,

The project is not dead 🎉 

I've had zero spare time for the past couple of months. I just working on big company and... just work.

I recently finished working on a `fast-glob` project that will allow to accelerate work with the directories several times. The `syncy` project right now has very high priority for me. You can check progress in `master` branch and #41 issue.

**Comment by fourthmeal70** [comment]
I did try changing the source as you suggested @eisbehr and it did return the run times back to normal! Thanks for that! 

@mrmlnc, Glad to hear this project isn't dead and is a high priority! Looking forward to what v3 will have to offer :)

**Comment by eisbehr-** [comment]
Glad it worked for you too! I still don't understand, why the destination files needs to be collected, if they are ignored?! 🤔

**Comment by fourthmeal70** [comment]
@eisbehr- Have you had a chance to try this again with the changes that have been made? The part of the code we changed has been rewritten to use fast-glob instead of globby. I'm wondering if it's made any difference. 

If you haven't tried it, I might try to give it a shot later this week or early next week.

**Comment by mrmlnc** [comment]
JFYI: version in the `master` branch is not released and  does not ready for users :)

**Comment by eisbehr-** [comment]
@fourthmeal70 Not yet, actually. I will give this a try, but it will take a while till I'm back on this project. In the next 2-3 month. ;)

**Comment by fourthmeal70** [comment]
Right, thanks for the info @mrmlnc . Maybe I'll wait until an official release then. I'd just like to get off the fork I created using the change mentioned in comments above. I had to force my development team to download and install Git so that their local builds could pull down the forked code. It wasn't a huge deal as we're working on moving all of our projects into Git anyway. But I'll be happier once I can remove that Github link from the package.json file :)

**Comment by jthannah** [comment]
I've installed Version 3 and tested out this issue and although better it's still not ideal. The copy takes about 30 minutes in my situation whereas with the change mentioned above the sync was happening in about 2 minutes. I'm still not sure why the paths I"m telling syncy to ignore in the destination have to be processed. It's seems like a waste of time and memory.
